packages/studio/src/components/AddPageButton/StreamScopeCollector.tsx

```diff
@@ -1,58 +1,185 @@
+        return numItems + scopeItems.length;
+      },
+      0
+    );
+
+    return (
+      <>
+        <div className="italic mb-4">
+          Use one of the optional fields below to specify which entities this
+          page can access.
+        </div>
+        {Object.entries(scopeFormData).map(([field, data]) => {
+          const selectedItems: string[] | undefined = streamScope?.[field];
+          const hasOtherScopeFilters =
+            totalStreamScopeItems > (selectedItems?.length ?? 0);
+          return (
+            <ScopeField
+              key={field}
+              field={field}
+              selectedItems={selectedItems}
+              updateScopeItems={updateScopeItems(field)}
+              disabled={data.allItems.length > 0 && hasOtherScopeFilters}
+              {...streamScopeFormData[field]}
+              {...data}
+            />
+          );
+        })}
+      </>
+    );
+  }, [scopeFormData, streamScope, updateScopeItems]);
+
   return (
-    <FormModal
+    <DialogModal
       isOpen={isOpen}
       title="Content Scope"
-      instructions="Use the optional fields below to specify which entities this page can access. Values should be separated by commas."
-      formData={streamScopeFormData}
-      closeOnConfirm={false}
-      confirmButtonText="Next"
       handleClose={handleClose}
-      handleConfirm={onConfirm}
+      handleConfirm={handleConfirm}
+      body={modalBodyContent}
+      confirmButtonText="Next"
+      isConfirmButtonDisabled={false}
     />
   );
 }
+
+interface ScopeFieldProps {
+  field: string;
+  description: string;
+  allItems: string[];
+  tooltip: string;
+  emptyText: string;
+  selectedItems: string[] | undefined;
+  updateScopeItems: (selectedItems: string[] | undefined) => void;
+  getDisplayName?: (item: string) => string;
+  disabled?: boolean;
+}
+
+function ScopeField({
+  field,
+  description,
+  allItems,
+  tooltip,
+  emptyText,
+  selectedItems,
+  updateScopeItems,
+  getDisplayName,
+  disabled,
+}: ScopeFieldProps) {
+  const addItem = useCallback(
+    (item: string) => {
+      updateScopeItems([...(selectedItems ?? []), item]);
+    },
+    [selectedItems, updateScopeItems]
+  );
+
+  const removeItem = useCallback(
+    (item: string) => {
+      updateScopeItems(selectedItems?.filter((i) => i !== item));
+    },
+    [selectedItems, updateScopeItems]
+  );
+
+  const availableItems = allItems.filter((i) => !selectedItems?.includes(i));
+  const labelId = `${field}-label`;
+
+  return (
+    <>
+      <label id={labelId}>{description}</label>
+      <Tooltip anchorSelect={`#${labelId}`} content={tooltip} />
+      <PillPickerInput
+        selectedItems={selectedItems}
+        availableItems={availableItems}
+        selectItem={addItem}
+        removeItem={removeItem}
+        getDisplayName={getDisplayName}
+        disabled={disabled}
+        customContainerClasses="mt-2 mb-4 border-gray-500"
+        emptyText={emptyText}
```

Comment:
Would it be possible to write this component in a way where it takes in less props?

Reply:
I gave it some thought and consolidated the logic to reduce a couple props, but I'm having a hard time reducing it any further. A couple things I considered but didn't end up doing:
- Moving the `empty` display case into the parent components (to remove `emptyText`): This is definitely doable, but the logic is pretty similar for both the Tailwind prop input and stream scope input, so moving it into the parents involved a little code duplication. Also, the `pillContainerClass` styling would need to be exported and used by both parent components, and they'd also have to handle adding the gray "disabled" background, which looked a little messy
- Moving the display name logic fully into `ScopeFilterField` (to remove `getDisplayName`): This is also possible, but we'd have to do some kind of regex matching or deconstructing of the display string to extract the `id` when updating the stream scope, which didn't seem ideal

Reply:
It looks a lot better!